src/queue.js

```diff
@@ -0,0 +1,217 @@
+var _ = require('./utility');
+
+/*
+ * Queue - an object which handles which handles a queue of items to be sent to Rollbar.
+ *   This object handles rate limiting via a passed in rate limiter, retries based on connection
+ *   errors, and filtering of items based on a set of configurable predicates. The communication to
+ *   the backend is performed via a given API object.
+ *
+ * @param rateLimiter - An object which conforms to the interface
+ *    rateLimiter.shouldSend(item) -> bool
+ * @param api - An object which conforms to the interface
+ *    api.postItem(payload, function(err, response))
+ * @param settings - see Queue.prototype.configure
+ */
+function Queue(rateLimiter, api, settings) {
+  this.rateLimiter = rateLimiter;
+  this.api = api;
+  this.settings = settings;
+  this.predicates = [];
+  this.pendingRequests = [];
+  this.retryQueue = [];
+  this.retryHandle = null;
+  this.waitCallback = null;
+}
+
+/*
+ * configure - updates the settings this queue uses
+ *
+ * @param settings - an object with the following possible keys
+ *    * something
+ *    * else
+ */
+Queue.prototype.configure = function(settings) {
+  var oldSettings = this.settings;
+  this.settings = _.extend(true, {}, oldSettings, settings);
+  return this;
+};
+
+/*
+ * addPredicate - adds a predicate to the end of the list of predicates for this queue
+ * 
+ * @param predicate - function(item, settings) -> (bool|{err: Error})
+ *  Returning true means that this predicate passes and the item is okay to go on the queue
+ *  Returning false means do not add the item to the queue, but it is not an error
+ *  Returning {err: Error} means do not add the item to the queue, and the given error explains why
+ *  Returning {err: undefined} is equivalent to returning true but don't do that
+ */
+Queue.prototype.addPredicate = function(predicate) {
+  if (_.isFunction(predicate)) {
+    this.predicates.push(predicate);
+  }
+  return this;
+};
+
+/*
+ * addItem - Send an item to the Rollbar API if all of the predicates are satisfied
+ *
+ * @param item - The payload to send to the backend
+ * @param callback - function(error, repsonse) which will be called with the response from the API
+ *  in the case of a success, otherwise response will be null and error will have a value. If both
+ *  error and response are null then the item was stopped by a predicate which did not consider this
+ *  to be an error condition, but nonetheless did not send the item to the API.
+ */
```

Comment:
I thought I was going to need `force` to allow some requests to be sent even if we are trying to drain the queue with wait, but that didn't turn out to be needed so I can get rid of it for the time being.